### PR TITLE
api: move HTTPProxy health check to route level

### PR DESCRIPTION
**Pull request by davecheney**
Updates #1119

This PR moves the HealthCheck stanza from the individual service level
to the route. This is an insurance policy if we have to move from one
cluster per service to using one service per route and multiple
clusterloadassignments. If that happens, we will not be able to support
per service configuration.

Specifically:

- Fork httpproxy/v1/healthCheck into ingressroute/v1beta/healthCheck
retaining the original feature set.
- Rename httpproxy/v1/healthCheck to healthCheckPolicy, making it
consistent with timeoutPolicy and retryPolicy.
- The underlying type is renamed to HTTPHeathCheckPolicy as it is
attached to Route.
- Healthchecks on httproxy.tcpproxy are removed. We can add them back in
rc.1 or rc.2.

Signed-off-by: Dave Cheney <dave@cheney.net>

### CODE REVIEW

**Comment by davecheney** [comment]
The intention of this PR is _if_ we don't need to move to one cluster per route, we can restore the health check per backend setting as an addition to the per route health check. The former, if present, will take precedence over the latter.

**Comment by stevesloka** [comment]
The original reason it was on the service was that each upstream could possibly have a different health check. For Gimbal, one service could be k8s and the other openstack. 

I know we struggled on the design originally if they would have different health checks or not in practice, so defaulted to allow them to have their own. 

TLDR: I think it's probably fine to have a single active healthcheck defined.

**Comment by stevesloka** [comment]
Something to think about for this design change might be the passive health checks, if we enabled those in the future, what might this health check policy look like? (Fair enough to call YAGNI as well).

**Comment by davecheney** [comment]
> The original reason it was on the service was that each upstream could possibly have a different health check. For Gimbal, one service could be k8s and the other openstack.
> 
> I know we struggled on the design originally if they would have different health checks or not in practice, so defaulted to allow them to have their own.
> 
> TLDR: I think it's probably fine to have a single active healthcheck defined.

Yeah, I'm not happy about rolling this back -- per service load balancing and health checks is very useful. 

The worry that I have is if we have to move to one cluster per route to resolve #1119 then that means having to deprecate the route.service.healthcheck field. Maybe we can table this change and look into #1119 some more. This doesn't need to happen this week, but I think we need to have a plan in place to move, or remove, the fields that would block us from moving to one cluster per route if required.

File: apis/projectcontour/v1/httpproxy.go [hunk 1]
```diff
@@ -105,23 +105,25 @@ type TLS struct {
...
-	// Indicates that during forwarding, the matched prefix (or path) should be swapped with this value
+	// Indicates that during forwarding, the matched prefix (or path) should be swapped with this value.
 	PrefixRewrite string `json:"prefixRewrite,omitempty"`
-	// The timeout policy for this route
+	// The timeout policy for this route.
 	TimeoutPolicy *TimeoutPolicy `json:"timeoutPolicy,omitempty"`
-	// The retry policy for this route
+	// The retry policy for this route.
 	RetryPolicy *RetryPolicy `json:"retryPolicy,omitempty"`
+	// The health check policy this route.
```
Comment (youngnick) [inline apis/projectcontour/v1/httpproxy.go]: Nit: 'for this route'.
